tests/platform/broadcom/files/ser_injector.py

```diff
@@ -0,0 +1,312 @@
+
+    def enable_sram_scan(self, interval_usec, rate):
+        '''
+        @summary: Enable Broadcom sram scan
+        @param interval_usec: memory scanner interval i usec
+        @param rate: rate (number of entries) per interval
+        '''
+        self.enable_memory_scan('sramscan', interval_usec, rate)
+
+    def verify_ser(self, entry, log):
+        '''
+        @summary: verify SER log entry
+        @param entry: indext of the memory table where SER was injected
+        @param log: syslog log line
+
+        @return: memory table name
+        @return: Flag if SER injection entry matches log line entry
+        '''
+        m = re.search("^.*addr:(.*) port.*index: (\d+)", log)
+
+        address = int(m.group(1), 16)
+        mem_entry = int(m.group(2))
+
+        memory = self.bcmMemory.get_memory_by_address()
+        if address in memory:
+            return memory[address], entry == mem_entry
+
+        return None, None
+
+    def inject_ser(self, mem, index = 0):
+        '''
+        @summary: Inject SER error suing Broadcom ser inject command
+        @param mem: name of the memory table to inject SER into
+        @param index: index of the entry to inject SER into
+        '''
+        return run_cmd(["bcmcmd",  "ser inject memory=" + mem + " index=" + str(index)])
+
+    def verify_and_update_test_result(self, entry, line):
+        '''
+        @summary: Verify log line and update test result
+        @param entry: index of the entry to inject SER into
+        @param log: syslog log line
+        '''
+        mem, entry_found = self.verify_ser(entry, line)
+        if mem is not None:
+            # memory could be aliased, mark all aliased memory as passed/failed
+            for m in mem:
+                if entry_found:
+                    if m in self.mem_verified:
+                        self.mem_verified[m] += 1
+                    else:
+                        print("Successfully tested memory %s" % m)
+                        self.mem_verified.update({m : 1})
+                else:
+                    if m in self.mem_failed:
+                        self.mem_failed[m] += 1
+                    else:
+                        print("Failed verification for memory %s, syslog '%s'" % (m, line))
+                        self.mem_failed.update({m : 1})
+
+                if m in self.mem_verification_pending:
+                    self.mem_verification_pending.remove(m)
+                else:
+                    print("Memory %s appeared more than once" % m)
+        else:
+            print("Memory corresponding to the following syslog was not found! Syslog: '%s'" % line)
+
+    def run_test(self, memory, entry = 0):
+        '''
+        @summary: Run SER injection test on cached memory tables
+        @param memory: Cached memory tables
+        @param entry: index of the entry to inject SER into
+        '''
+        with open('/var/log/syslog') as syslog_file:
```

Comment:
What if /var/log/syslog is rotated?

Reply:
The test has some resiliency built into it as it runs on cache-able/testable memory tables 3 times (configurable).  When a memory table is tested successfully, it is removed from the pending verification list. Next run takes place on those pending verification memory tables list. This behavior will compensate for losses happening in syslog logging system and also for losses due to log rotation. I also found out that logs are rotated weekly.

Reply:
what if disable rotation for the test?

Reply:
That could be done. Syslog log is lossy so the above iterative approach is build to handle such losses in loglines. It will handle the log rotation as well. The extreme case is that log get rotated right after the reading phase of the nth iteration which will render iteration n fruitless. Remaining iteration will not suffer from log rotation and the test should finish.